src/treetensornetworks/projttns/abstractprojttn.jl

```diff
@@ -121,25 +126,55 @@ function Base.size(P::AbstractProjTTN)::Tuple{Int,Int}
     end
   end
   for j in sites(P)
-    for i in inds(P.H[j])
+    for i in inds(operator(P)[j])
       plev(i) > 0 && (d *= dim(i))
     end
   end
   return (d, d)
 end
 
 function position(
-  P::AbstractProjTTN{V}, psi::TTN{V}, pos::Union{Vector{<:V},NamedEdge{V}}
+  P::AbstractProjTTN{V}, psi::AbstractTTN{V}, pos::Union{Vector{<:V},NamedEdge{V}}
+) where {V}
+  P = shift_position(P, pos)
+  P = invalidate_environments(P)
+  P = make_environments(P, psi)
+  return P
+end
+
+function position!(
+  P::AbstractProjTTN{V}, psi::AbstractTTN{V}, pos::Union{Vector{<:V},NamedEdge{V}}
 ) where {V}
-  # shift position
   P = shift_position(P, pos)
-  # invalidate environments corresponding to internal edges
   for e in internal_edges(P)
     unset!(P.environments, e)
   end
-  # make all environments surrounding new position
+  make_environments!(P, psi)
+  return P
+end
+
+function invalidate_environments(P::AbstractProjTTN)
+  ie = internal_edges(P)
+  newenvskeys = filter(!in(ie), keys(P.environments))
+  P = set_environments(P, getindices(P.environments, newenvskeys))
+  return P
+end
+
+function invalidate_environment(P::AbstractProjTTN, e::NamedEdge)
```

Comment:
```suggestion
function invalidate_environment(P::AbstractProjTTN, e::AbstractEdge)
```
to be more agnostic about what kind of graph it may be.